### PR TITLE
AG-17220 fix inner header component destroyed on every column def update

### PR DESCRIPTION
**Pull request by StephenCooper**
## Problem

`HeaderComp.refresh()` compared the current template against the result of `workOutTemplate()`, but that function returns a **new** `ElementParams` object on every call. The reference was never equal, so `refresh()` always returned `false`, triggering a full destroy + remount of the header cell (and its `innerHeaderComponent`) on every `setColumnDefs` call.

A second, related issue: when `innerHeaderComponent.refresh()` returns `false` (signalling it cannot handle the update), the return value was silently ignored — the inner component was left stale. Worse, the outer `HeaderComp` would have been remounted unnecessarily if it returned false because of the template comparison bug above.

## Fix

- Replace the single template-reference comparison with three explicit boolean checks: custom template string, formula-active state, and sort-indicator presence. These are exactly the conditions `workOutTemplate()` inspects, so the equality check is now direct and allocation-free.
- When `innerHeaderComponent.refresh()` returns `false`, only the inner component is replaced (`replaceInnerHeaderComponent`) — the outer `HeaderComp` stays alive.
- Reorder `replaceInnerHeaderComponent` to detach the old GUI from the DOM before calling `destroyBean`, so the element is not briefly orphaned in `eText`.

## Tests

Two new behavioural test files:

- `columns-header-inner.test.ts` — vanilla-JS grid, verifies both scenarios (refresh not remount; inner component remount when refresh returns false, outer header unaffected).
- `columns-header-inner.test.tsx` — React grid, same two scenarios using `useEffect` mount/unmount counters.

Fix [AG-17220](https://ag-grid.atlassian.net/browse/AG-17220)

### CODE REVIEW

File: packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts [hunk 1]
```diff
@@ -112,7 +123,9 @@ export class HeaderComp extends Component implements IHeaderComp {
             // Mimic the merging of params that happens during init of _getInnerHeaderCompDetails(userCompFactory, params, params);
             const mergedParams = { ...params };
```
[the remark this answers is not in the document]
Comment (StephenCooper) [reply packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts:104]: Addressed

[the remark this answers is not in the document]
Comment (StephenCooper) [reply packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts:104]: addressed.

File: packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts [hunk 2]
```diff
@@ -79,31 +78,44 @@ export class HeaderComp extends Component implements IHeaderComp {
...
     public refresh(params: IHeaderParams): boolean {
-        const oldParams = this.params;
+        const {
+            beans: { formula, sortSvc },
+            currentTemplateCustom,
+            currentFormulaActive,
+            currentShowMenu,
+            currentHasSortIndicator,
+            currentSort,
+            currentSuppressMenuHide,
```
[the remark this answers is not in the document]
Comment (StephenCooper) [reply packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts]: Addressed even though it was a non issue as the selector is hardcoded in that method.

File: packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts [hunk 3]
```diff
@@ -112,7 +124,12 @@ export class HeaderComp extends Component implements IHeaderComp {
             // Mimic the merging of params that happens during init of _getInnerHeaderCompDetails(userCompFactory, params, params);
             const mergedParams = { ...params };
             _mergeDeep(mergedParams, params.innerHeaderComponentParams);
-            this.innerHeaderComponent.refresh?.(mergedParams);
+            if (
+                params.innerHeaderComponent !== this.currentInnerHeaderComponent ||
```
[the remark this answers is not in the document]
Comment (StephenCooper) [reply packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts:107]: Fixed

File: packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts [hunk 4]
```diff
@@ -112,7 +124,12 @@ export class HeaderComp extends Component implements IHeaderComp {
...
+                !this.innerHeaderComponent.refresh?.(mergedParams)
```
[the remark this answers is not in the document]
Comment (StephenCooper) [reply packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts]: While technically correct I think this is being over defensive and not worth the added complexity and code bloat.

File: packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts [hunk 5]
```diff
@@ -79,31 +78,44 @@ export class HeaderComp extends Component implements IHeaderComp {
...
+            !!formula?.active !== currentFormulaActive ||
+            !!sortSvc?.getSortIndicatorSelector() !== currentHasSortIndicator ||
+            params.enableSorting != currentSort ||
             oldParams.enableFilterButton != params.enableFilterButton ||
-            oldParams.enableFilterIcon != params.enableFilterIcon
+            oldParams.enableFilterIcon != params.enableFilterIcon ||
+            (params.column as AgColumn).formulaRef != this.currentRef ||
+            (params.template?.trim?.() ?? params.template) !== currentTemplateCustom ||
+            this.workOutShowMenu() != currentShowMenu ||
+            (currentSuppressMenuHide != null && this.shouldSuppressMenuHide() != currentSuppressMenuHide)
```
Comment (sdwvit) [inline packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts]: can you please simplify this condition, also a lot of implicit casting with `!=` may benefit from narrowing

File: packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts (hunk 4 above)
Comment (sdwvit) [inline packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts]: does it have a side effect? if so pls extract it into a variable

File: packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts [hunk 6]
```diff
@@ -133,10 +150,13 @@ export class HeaderComp extends Component implements IHeaderComp {
     public init(params: IHeaderParams): void {
         this.params = params;
 
-        const { sortSvc, touchSvc, rowNumbersSvc, userCompFactory } = this.beans;
+        const { sortSvc, touchSvc, rowNumbersSvc, formula } = this.beans;
         const sortComp = sortSvc?.getSortIndicatorSelector();
-        this.currentTemplate = this.workOutTemplate(params, !!sortComp);
-        this.setTemplate(this.currentTemplate, sortComp ? [sortComp] : undefined);
+        const template = this.workOutTemplate(params, !!sortComp);
+        this.currentTemplateCustom = params.template?.trim?.() ?? params.template;
```
Comment (sdwvit) [inline packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts]:
```suggestion
        this.currentTemplateCustom = params.template?.trim?.();
```
should be enough?

File: packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts [hunk 7]
```diff
@@ -133,10 +150,13 @@ export class HeaderComp extends Component implements IHeaderComp {
...
+        this.currentFormulaActive = !!formula?.active;
+        this.currentHasSortIndicator = !!sortComp;
+        this.setTemplate(template, sortComp ? [sortComp] : undefined);
```
Comment (sdwvit) [inline packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts:160]: why there needs to be an array?

File: packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts [hunk 8]
```diff
@@ -79,31 +78,44 @@ export class HeaderComp extends Component implements IHeaderComp {
...
+    private currentInnerHeaderComponent: IHeaderParams['innerHeaderComponent'];
     private isLoadingInnerComponent: boolean = false;
 
     private mouseListener?: HeaderCellMouseListenerFeature;
 
     public refresh(params: IHeaderParams): boolean {
-        const oldParams = this.params;
+        const {
+            beans: { formula, sortSvc },
+            currentTemplateCustom,
```
Comment (sdwvit) [inline packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts]: more an arch question, but why do we need to keep a 'current' state in this comp?

File: packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts [hunk 9]
```diff
@@ -174,15 +196,25 @@ export class HeaderComp extends Component implements IHeaderComp {
 
             if (this.isAlive()) {
                 this.innerHeaderComponent = comp;
-                if (this.eText) {
-                    this.eText.appendChild(comp.getGui());
-                }
+                this.eText?.appendChild(comp.getGui());
```
Comment (sdwvit) [inline packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts:202]: what happens if eText is null

File: packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts [hunk 10]
```diff
@@ -174,15 +196,25 @@ export class HeaderComp extends Component implements IHeaderComp {
...
             } else {
                 this.destroyBean(comp);
             }
         });
     }
 
+    private replaceInnerHeaderComponent(params: IHeaderParams): void {
+        const oldGui = this.innerHeaderComponent!.getGui();
```
Comment (sdwvit) [inline packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts]: either the type needs refinement or ! assertion is not needed here

File: packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts (hunk 6 above)
Comment (SalvatorePreviti) [reply packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts]: why do we need trim though?

**Review by SalvatorePreviti — APPROVED** [review]

File: packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts (hunk 4 above)
Comment (StephenCooper) [reply packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts]: To extract I would need to have a nested if statement which I don't think would work out as clean. We don't want to call refresh on the old component if it is different to the new inner component.

However, this has highlighted a missing test case. A column that does not have an innerHeaderComponent but then is updated to have one.

File: packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts (hunk 6 above)
Comment (StephenCooper) [reply packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts]: I will simplify to just store params.template for tracking changes. The trim is only needed when we go to render it.

File: packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts (hunk 7 above)
Comment (StephenCooper) [reply packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts:160]: Its what setTemplate expects.

File: packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts (hunk 9 above)
Comment (StephenCooper) [reply packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts:202]: The same behaviour as currently. I have just inlined the optional chaining.

File: packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts (hunk 10 above)
Comment (StephenCooper) [reply packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts]: Will be fixed as part of the missing test case.

File: packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts (hunk 8 above)
Comment (StephenCooper) [reply packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts]: Previous implementation comment is that it was more effort to perform the selective updates then just rebuild when any part changes.

File: packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts (hunk 9 above)
Comment (sdwvit) [reply packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts:202]: if eText is null appendChild would resolve into an undefined and throw in runtime, I was referring to this :)

File: packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts (hunk 5 above)
Comment (StephenCooper) [reply packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts]: Have simplified and pulled into its own method.

File: packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts [hunk 11]
```diff
@@ -79,13 +78,13 @@ export class HeaderComp extends Component implements IHeaderComp {
     public params: IHeaderParams;
 
     private currentDisplayName: string;
-    private currentTemplate: ElementParams | string | null | undefined;
+    private currentFormulaActive: boolean | undefined;
     private currentShowMenu: boolean;
     private currentSuppressMenuHide: boolean;
-    private currentSort: boolean | undefined;
```
Comment (StephenCooper) [inline packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts:85]: This cannot change as it is recording the inclusion of the sortService. Modules cannot dynamically change

**Review by SalvatorePreviti — APPROVED** [review]

File: testing/behavioural/src/columns/headers/columns-header-inner.test.ts [hunk 12]
```diff
@@ -0,0 +1,549 @@
...
+        (TrackedComp.prototype as any).refresh = function (_params: IHeaderParams): boolean {
+            TrackedComp.refreshCount++;
+            return true;
+        };
+    } else if (refreshBehaviour === 'reject') {
+        (TrackedComp.prototype as any).refresh = function (_params: IHeaderParams): boolean {
+            TrackedComp.refreshCount++;
+            return false;
+        };
+    }
```
[the remark this answers is not in the document]
Comment (SalvatorePreviti) [reply testing/behavioural/src/columns/headers/columns-header-inner.test.ts]: mhh yes this is a bit weird, there is not a better way? fine with me though but maybe there is an easier / cleaner way

File: packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts [hunk 13]
```diff
@@ -174,15 +196,27 @@ export class HeaderComp extends Component implements IHeaderComp {
...
             }
         });
     }
 
+    private replaceInnerHeaderComponent(params: IHeaderParams): void {
+        const curr = this.innerHeaderComponent;
+        if (curr) {
+            _removeFromParent(curr.getGui());
+            this.destroyBean(curr);
+        }
```
[the remark this answers is not in the document]
Comment (SalvatorePreviti) [reply packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts:215]: to verify

File: packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts (hunk 9 above)
Comment (StephenCooper) [reply packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts:202]: This is not a real issue. If eText is null the full statement is skipped. appendChild is always present if eText is defined. That would be when the second one is required.

File: packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (StephenCooper) [reply packages/ag-grid-community/src/headerRendering/cells/column/headerComp.ts:215]: Was correct, extended tests to cover this edge case.

File: testing/behavioural/src/columns/headers/columns-header-inner.test.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (StephenCooper) [reply testing/behavioural/src/columns/headers/columns-header-inner.test.ts]: Cleaned up

**Comment by StephenCooper** [comment]
I am going to close this PR for now as the scope has grown too much for the severity of the original issue.
